x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/tree/Node.java

```diff
@@ -81,7 +85,11 @@ public <E extends T> void forEachDown(Class<E> typeToken, Consumer<? super E> ac
 
     @SuppressWarnings("unchecked")
     public void forEachUp(Consumer<? super T> action) {
-        children().forEach(c -> c.forEachUp(action));
+        // please do not refactor it to a for-each loop
+        // to avoid allocating iterator that performs concurrent modification checks
+        for (int c = 0; c < children.size(); c++) {
+            children.get(c).forEachUp(action);
+        }
```

Comment:
forEachDown and forEachUp do not seem to have visible impact on profile I reviewed, however replacing them to loops should make profiles outputs several stacks shorter, especially when those are nested.

![Screenshot at 2025-03-19 15-21-23](https://github.com/user-attachments/assets/48a593be-b4c0-485c-a010-e63307959100)